HW2/P2/parallel_vector.pyx

```diff
@@ -99,17 +111,28 @@ cpdef move_data_medium_grained(np.int32_t[:] counts,
        int num_locks = (counts.shape[0] + N - 1) / N  # ensure enough locks
        omp_lock_t *locks = get_N_locks(num_locks)
 
-   ##########
-   # Your code here
-   # Use parallel.prange() and a lock for every N adjacent elements of counts
-   # to parallelize data movement.  Be sure to avoid deadlock, as well as
-   # double-locking.
-   ##########
-   with nogil:
+   with nogil, parallel(num_threads = 4):
        for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
+           for idx in prange(src.shape[0]):
+           # Case 1: when source and destination locations are different
+               if( src[idx]/N != dest[idx]/N ):   
```

Comment:
Usually, you want to do a check again after acquiring the lock, since the condition might have changed between checking the condition and acquiring the lock